### PR TITLE
Add content negotiation support and sharding parameters for Query Frontend

### PR DESCRIPTION
**Pull request by annanay25**
**What this PR does**:
This PR is the first half the work to introduce a Query Frontend (follow up of #362), and does a bunch of things:
- Add parameters to HTTP endpoint of the Querier, which will help with sharding. List of parameters -
  - **BlockStart & BlockEnd** - to specify the shard space for each querier
  - **QueryIngesters** - to specify whether a querier should query ingesters (only one of the queriers should query ingesters for every query)
- Add content negotiation between Tempo-query and Tempo-querier to marshal trace data into proto (and save a roundtrip b/n proto & json)

**Which issue(s) this PR fixes**:
Fixes #NA

**Checklist**
- [x] Tests updated
- NA Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -83,7 +83,7 @@ lint:
 
 .PHONY: docker-component # Not intended to be used directly
 docker-component: check-component exe
-	docker build -t grafana/$(COMPONENT) --platform=linux/$(GOARCH) --build-arg=TARGETARCH=$(GOARCH) -f ./cmd/$(COMPONENT)/Dockerfile .
```
Comment (annanay25) [inline Makefile:86]: Oops, moving this to a separate PR.

File: modules/querier/http.go [hunk 2]
```diff
@@ -50,6 +65,21 @@ func (q *Querier) TraceByIDHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if r.Header.Get("Tempo-query-content-type") == "application/grpc" {
```
Comment (annanay25) [inline modules/querier/http.go:68]: This block can be dropped.

**Comment by joe-elliott** [comment]
So there's like 3 things going on here

1) Additional parameters on the query endpoint
2) Content negotiation to proto for the HTTP endpoint
3) GRPC for the querier

Given that we know the query frontend is going to want to use HTTP I think it makes sense to drop #3 and only do #1 and #2.  Thoughts?

File: modules/querier/http.go (hunk 2 above)
Comment (joe-elliott) [reply modules/querier/http.go:68]: There's a standard header to use here:  "Accept".

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

As far as the MIME type it looks like there's some discussion:

https://stackoverflow.com/questions/30505408/what-is-the-correct-protobuf-content-type

`application/protobuf` is fine

**Comment by annanay25** [comment]
Hmm, Frontend is going to roundtrip GRPC to the Queriers no? It's going to reuse this - https://github.com/cortexproject/cortex/blob/master/pkg/frontend/v2/frontendv2pb/frontend.proto

**Comment by joe-elliott** [comment]
I don't think the query-frontend supports GRPC.  Isn't it all built around that http wrapped in grpc trick?  Don't we need http on the queriers to use it without major changes?

**Comment by annanay25** [comment]
We can add a GRPC handler to the query frontend. QF <> Querier will be GRPC bi-directional stream. Just the payload is HTTP, but we can marshal proto into the body like we're doing between QF and Jaeger-GRPC plugin.

File: cmd/tempo-query/tempo/plugin.go [hunk 3]
```diff
@@ -59,6 +59,9 @@ func (b *Backend) GetTrace(ctx context.Context, traceID jaeger.TraceID) (*jaeger
 		req.Header.Set(user.OrgIDHeaderName, tenantID)
 	}
 
+	// Set content type to grpc
+	req.Header.Set("Accepts", "application/protobuf")
```
Comment (joe-elliott) [inline cmd/tempo-query/tempo/plugin.go:63]: Let's use the standard header `Accept`:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

Probably would be wise to define this in one place and reference it everywhere we need it.  Perhaps a new file: `util/http/headers.go` ?  I'm also surprised to see I'm not finding this defined in the go std lib.

File: modules/querier/http.go [hunk 4]
```diff
@@ -50,6 +57,21 @@ func (q *Querier) TraceByIDHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if r.Header.Get("Accepts") == "application/protobuf" {
```
Comment (joe-elliott) [inline modules/querier/http.go:60]: same as above

File: modules/querier/http.go [hunk 5]
```diff
@@ -50,6 +57,21 @@ func (q *Querier) TraceByIDHandler(w http.ResponseWriter, r *http.Request) {
...
+		level.Info(cortex_util.Logger).Log("msg", "received content type application/grpc")
```
Comment (joe-elliott) [inline modules/querier/http.go:61]: i don't think we want to log this everytime.  it will just become spam in the logs

File: modules/querier/http.go [hunk 6]
```diff
@@ -50,6 +57,21 @@ func (q *Querier) TraceByIDHandler(w http.ResponseWriter, r *http.Request) {
...
+		trace := &tempopb.Trace{Batches: resp.Trace.Batches}
```
Comment (joe-elliott) [inline modules/querier/http.go:62]: is there a need to create a new trace here?  `resp.Trace` should be marshallable directly?

File: modules/querier/http.go [hunk 7]
```diff
@@ -37,9 +42,11 @@ func (q *Querier) TraceByIDHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	resp, err := q.FindTraceByID(ctx, &tempopb.TraceByIDRequest{
-		TraceID: byteID,
+		TraceID:        byteID,
```
Comment (joe-elliott) [inline modules/querier/http.go:45]: we need to handle the case where none of these parameters are present.  default block start to `0x00...`, end to `0xFF...` and query ingesters to true.

also, how are you encoding/decoding blockstart/end key?  i don't think a direct cast of a string to bytes is what we want.  probably need to encode/decode a hex string.

we should also parse the parameters here and return an appropriate error if any parameters don't make sense.  i.e. a blockstart/end string that is not hex encoded or a query ingesters param is not "true" or "false".

File: pkg/tempopb/tempo.pb.go [hunk 8]
```diff
@@ -28,7 +28,10 @@ var _ = math.Inf
 const _ = proto.GoGoProtoPackageIsVersion3 // please upgrade the proto package
 
 type TraceByIDRequest struct {
-	TraceID []byte `protobuf:"bytes,1,opt,name=traceID,proto3" json:"traceID,omitempty"`
+	TraceID        []byte `protobuf:"bytes,1,opt,name=traceID,proto3" json:"traceID,omitempty"`
+	BlockStart     []byte `protobuf:"bytes,2,opt,name=blockStart,proto3" json:"blockStart,omitempty"`
```
Comment (joe-elliott) [inline pkg/tempopb/tempo.pb.go:32]: block start/end are guids.  i think we should just use strings b/c it would make things easier to debug and easier to manually query.  this impacts my comment above about choosing default params.  i suggested 0x00/0xFF but it would be a guid string instead that we'd default to.

File: tempodb/tempodb.go [hunk 9]
```diff
@@ -79,7 +79,7 @@ type Writer interface {
 }
 
 type Reader interface {
-	Find(ctx context.Context, tenantID string, id encoding.ID) ([]byte, FindMetrics, error)
+	Find(ctx context.Context, tenantID string, id encoding.ID, blockStart []byte, blockEnd []byte) ([]byte, FindMetrics, error)
```
Comment (joe-elliott) [inline tempodb/tempodb.go:82]: take uuid.UUID's instead of []bytes for blockStart/End

File: tempodb/tempodb.go [hunk 10]
```diff
@@ -246,14 +246,28 @@ func (rw *readerWriter) Find(ctx context.Context, tenantID string, id encoding.I
 	rw.blockListsMtx.Lock()
 	blocklist, found := rw.blockLists[tenantID]
 	copiedBlocklist := make([]interface{}, 0, len(blocklist))
+
+	var err error
```
Comment (joe-elliott) [inline tempodb/tempodb.go:250]: this code becomes much simpler if take a uuid.UUID directly.  no reason to double check if there's a value, just always respect the values passed in.

**Review by joe-elliott — CHANGES_REQUESTED** [review]
Some clean up requested.  This is definitely on the right path!

File: cmd/tempo-query/tempo/plugin.go (hunk 3 above)
Comment (annanay25) [reply cmd/tempo-query/tempo/plugin.go:63]: Agreed, added separate util file `http.go`.

File: modules/querier/http.go (hunk 5 above)
Comment (annanay25) [reply modules/querier/http.go:61]: This will be removed before merging. Add for debugging.

File: modules/querier/http.go (hunk 6 above)
Comment (annanay25) [reply modules/querier/http.go:62]: I swear I was seeing some issues with it being a pointer, but you're right, its marshallable directly.

File: modules/querier/http.go (hunk 7 above)
Comment (annanay25) [reply modules/querier/http.go:45]: Yeah, there are two approaches to this - one is to set/reset the query parameters at the http handler level (point of entry at modules/querier/http.go) and the second is to propagate them to the lowest level - the case where none of the parameters are present is handled here - https://github.com/grafana/tempo/pull/375/files#diff-d68ed2a23a87b802e3d85870a36046be8d2f61dd2caeab632503164c9b0ab57dR251

I agree the first approach is cleaner, and we can make it generic enough to be a validation layer for the parameters.

encoding/decoding is chosen as bytes for now because it is easier to split the block range when working with raw bytes. I'm open to changing it to a guid.String() which is hex encoded. The flow at the query frontend would then be - a) generate the block shard ranges using raw bytes b) perform the following uuid.FromBytes(blockRange).String() to generate the hex encoding

File: tempodb/tempodb.go (hunk 9 above)
Comment (annanay25) [reply tempodb/tempodb.go]: I changed this to string, because otherwise we would have to pass `uuid` in the grpc message `tempopb.TraceByIDRequest`, which doesn't exist yet. Or, we can change signature for the `FindTraceByID()` function.

Comment (annanay25) [reply tempodb/tempodb.go]: I think string is nicer, it adds another layer of checks in `tempodb.go` but we should anyway have checks there because we're trying to keep it independent of the rest of the codebase (modules and stuff).

But if we have strong opinions to pass uuid I can see what we can do to change to that.

File: tempodb/tempodb.go [hunk 11]
```diff
@@ -243,17 +248,41 @@ func (rw *readerWriter) Find(ctx context.Context, tenantID string, id encoding.I
...
 	copiedBlocklist := make([]interface{}, 0, len(blocklist))
+
 	for _, b := range blocklist {
 		// if in range copy
 		if bytes.Compare(id, b.MinID) != -1 && bytes.Compare(id, b.MaxID) != 1 {
-			copiedBlocklist = append(copiedBlocklist, b)
+			blockIDBytes, _ := b.BlockID.MarshalBinary()
+			// check block is in shard boundaries
+			// blockStartBytes <= blockIDBytes < blockEndBytes
+			if (bytes.Compare(blockIDBytes, blockStartBytes) >= 0) && (bytes.Compare(blockIDBytes, blockEndBytes) < 0) {
```
Comment (joe-elliott) [inline tempodb/tempodb.go:278]: this needs to be inclusive on both sides.  otherwise you won't ever search a block with the guid BlockIDMax.

File: pkg/util/http.go [hunk 12]
```diff
@@ -0,0 +1,6 @@
+package util
+
+const (
+	ContentTypeHeaderKey    = "Accept"
```
Comment (joe-elliott) [inline pkg/util/http.go:4]: call this `AcceptHeaderKey`.  There is another standard header called "ContentType"

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type

File: modules/querier/http.go [hunk 13]
```diff
@@ -36,10 +46,24 @@ func (q *Querier) TraceByIDHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// validate request
+	valid, blockStart, blockEnd, queryIngesters := validateAndSanitizeRequest(r)
+	if !valid {
+		http.Error(w, "invalid parameters", http.StatusBadRequest)
```
Comment (joe-elliott) [inline modules/querier/http.go:52]: it would be nice if `validateAndSanitizeRequest` returned an `error` with the reason for the failed validation which we then returned in the body.

**Review by joe-elliott — CHANGES_REQUESTED** [review]
A few small changes.  Looking very good.

**Comment by annanay25** [comment]
Addressed comments. @joe-elliott - This should be ready for review.

**Review by joe-elliott — APPROVED** [review]
